src/arch/50-abi.md

```diff
@@ -0,0 +1,707 @@
+
+If a user does not want to sign a message, bit `0` should be placed to the root cell start and signature omitted.
+
+The message body signature is generated from the *representation hash* of the bag of cells following the signature.
+
+## Signing Algorithm
+
+1. ABI serialization generates bag of cells containing header parameters, function ID and function parameters.
+   591 free bits are reserved in the root cell for destination address (the maximum size of internal address).
+2. The root cell data is prepended with actual destination address data without padding to maximum size.
+3. *Representation hash* of the bag is signed using the *Ed25519* algorithm.
+4. Address data is removed from the root cell and replaced with bit `1` followed by 512 bits of the signature.
+
+:::note
+This functionality is added since `ABI v2.3` and supported staring with [0.64.0](https://github.com/tonlabs/TON-Solidity-Compiler/blob/master/Changelog_TON.md#0640-2022-08-18) version of the Solidity compiler.
+:::
+
+## ABI JSON
+
+This section describes schema of the smart contracts ABI represented in JSON format.
+
+Full ABI schema in TypeScript notation:
+
+```typescript
+type Abi = {
+  version: string,
+  setTime?: boolean,
+  header?: Param[],
+  functions: Function[],
+  events?: Event[],
+  data?: Data[],
+  fields?: Param[],
+}
+
+type Function = {
+  name: string,
+  inputs?: Param[],
+  outputs?: Param[],
+  id?: number,
+}
+
+type Event = {
+  name: string,
+  inputs?: Param[],
+  id?: number,
+}
+
+type Data = Param & {
+  key: number,
+}
+
+type Param = {
+  name: string,
+  type: string,
+  components?: Param[],
+}
+```
+
+Where:
+
+- `version` contains string and uses semver semantics. Current version is "2.1".
+- `functions` describes all functions the smart contract can handle.
+- `events` describes all external outbound messages (events) produces by smart contract.
+- `data` describes Hashmap with public data of the smart contract.
+- `fields` describes internal structure of the smart contracts data.
+
+## Function Signature (Function ID)
+
+The following syntax is used for defining a signature:
+
+- function name
+- list of input parameter types (input list) in parenthesis
+- list of return values types (output list) in parenthesis
+- ABI version
+
+Single comma is used to divide each input parameter and return value type from one another. Spaces are not used.
+
+Parameter and return value names are not included.
+
+The function name, input and output lists are not separated and immediately follow each other.
+
+If a function has no input parameters or does not return any values, the corresponding input or output lists are empty (empty parenthesis).
+
+### Function Signature Syntax
+
+`function_name(input_type1,input_type2,...,input_typeN)(output_type1,output_type2,...,output_typeM)v2`
+
+### Signature Calculation Syntax
+
+`SHA256("function_name(input_type1,input_type2,...,input_typeN)(output_type1,output_type2,...,output_typeM)")v2`
```

Comment:
FIXME to `SHA256("function_name(input_type1,input_type2,...,input_typeN)(output_type1,output_type2,...,output_typeM)v2")`